### ISSUE TITLE
Keep copyright translations in config.toml

### ISSUE THREAD

**Issue by ZzMzaw** [welpo/tabi#214]
# Feature Request

## Summary
copyright is defined in config.toml but translations must be in i18n files.
It could be great to have both copyright and translations defined next to each other in the config.tom file.

## Motivation
Having translations next to the original copyright makes it easier to maintain.
In addition, it avoid having to duplicate i18n files in the website and so having to maintain them as well in case of tabi evolution just for one single key which is not linked with tabi translation.

## Detailed Description
I would keep using `translate_copyright` as copyright translation intent for backward compatibility.
I would introduce an array of translations (one per lang) defined as follow in config.toml extra section:
```toml
copyright_translations.es = "$TITLE © $CURRENT_YEAR Your Name $SEPARATOR A menos que se indique lo contrario, el contenido de este sitio web está disponible bajo la licencia [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)."
```

## Additional Context
I already have some code ready but it is mixed up with another change which could be interest for you (adding a $TITLE variable to copyright for config.title).
If you're ok with it, I can open a PR for you to look at it.

**Comment by welpo** [comment]
Lovely idea. I was never a fan of how the copyright being in the toml files.

> I would keep using `translate_copyright` as copyright translation intent for backward compatibility.

Well thought. Perhaps the logic could be: "try to find translation in toml file; if it's empty, try the new array".

I'd appreciate a PR for this.

---

> another change which could be interest for you (adding a $TITLE variable to copyright for config.title)

Sure, go ahead. Please make it a separate PR, though.

In both cases, we'll need to update the documentation (comments in config/theme.toml + the [Mastering tabi Settings](https://welpo.github.io/tabi/blog/mastering-tabi-settings/) post).

I can help with the Spanish and Catalan translations of the docs, of course.

If you need any help, don't hesitate to reach out.

Thanks!

**Comment by ZzMzaw** [comment]
Many thanks for your feedback.

> Perhaps the logic could be: "try to find translation in toml file; if it's empty, try the new array".

I don't know how you usually manage breaking changes but I would introduce a small one here to completely remove the use of i18n files for copyright.
The way I implemented the change will raise an error at build time if you were using the previous translation method and forgot to move translations back into `config.toml`. This would prevent people from silently falling back to the default copyright.

I let you have a look at the PR (#215) and tell me what you prefer to do.
It is completely possible to continue using the i18n files but I would use them as a fallback in case the translation array is empty (instead of using them as a primary) in order to give precedence to the new feature.

> Sure, go ahead. Please make it a separate PR, though.

It is really complex to have two different PR in parallel because the two changes both impact the same piece of code.
So far, I have two commits so you can easily look at each of them individually.
Once discussion on the translation will be completed, I'll remove the last commit and open a second PR for '$TITLE' if you prefer.

> we'll need to update the documentation

I had thought about the post but completely forgot about config/theme.toml.
Thanks for head's up, everything has been updated in the PR.

> I can help with the Spanish and Catalan translations of the docs

Many thanks. I can manage French and English but anything else go through online translation mixed with what you already wrote in the posts.
I would really appreciate if you could fix those approximate translations.

Thanks!

**Comment by welpo** [comment]
Great job! Thanks for the contribution.

I've pushed a commit with some code and docs changes.

> I don't know how you usually manage breaking changes but I would introduce a small one here to completely remove the use of i18n files for copyright.

I avoid them. Number 1 priority is to keep things stable. What I've done is:
- remove the `translate_copyright` variable from config and theme.toml.
- look for translations with the new variable
- if `copyright_translations` is not set and `translate_copyright = true`, we use the old code
- additionally, I've added a commented out error for when we remove the old way, in the future with proper notice (this will happen once I get to using versions for tabi + a changelog)

In short: new users will only see the new way. People using the old config can keep everything as is, or move to the new config, without any breaking changes.

>  I would really appreciate if you could fix those approximate translations.

Done!

>  Once discussion on the translation will be completed, I'll remove the last commit and open a second PR for '$TITLE' if you prefer.

It's fine. We'll merge it all together; I've modified the docs to reflect all the changes.

> I might not be able to connect for a couple days and I have a very irregular capacity to work on such personal topics so no worries if don't answer back quickly

Totally understandable. No worries at all.

By the way, I also added the variable `$AUTHOR` for the copyright notice, which gets replaced by the config's `author`.

Looking forward to your thoughts!

Thanks again.

**Comment by ZzMzaw** [comment]
Awesome.
I really appreciate the way you made it straightforward and future-proof.

> I avoid them. Number 1 priority is to keep things stable.

Go it, I'll keep it in mind for future contribution.

> By the way, I also added the variable $AUTHOR for the copyright notice, which gets replaced by the config's author.

Thanks to that (because author was not set on my side), I identified that both title and author are optional in config.toml and proposed a commit to cover case where it's not configured.

I also moved the PR to ready to review so you can merge it if it is fine with you

**Comment by welpo** [comment]
> proposed a commit to cover case where it's not configured

I think it's better to fail loudly if a user adds `$AUTHOR` to the copyright string, but hasn't set up `config.author`, no?

Otherwise, they might be wondering why there's nothing at all where they expect their name to be. I don't think failing silently (as per the last commit) is the best approach.

**Comment by welpo** [comment]
Ah, I see the problem. Zola complains even `$AUTHOR` or the other variables aren't set.

I'll think about whether there's a better option than what you've done (good catch).

**Comment by welpo** [comment]
Right, so I've:
- Moved the copyright code to a partial file (easier to read)
-Added a block to check if a $VARIABLE is set in the copyright but not set up in config. If that happens, it throws a descriptive error like:

```
Error: Reason: Function call 'throw' failed
Error: Reason: ERROR: The following variables are included in `copyright` but have not been set in the config.toml: author
```

I think it's ready for merge, but I'd love a final review from you in case I missed something (I often do) (-:

**Comment by ZzMzaw** [comment]
You exactly figured out what the issue was. I'll try to make my words more explicit in the future.

I completely agree it is far better to fail loudly if a used variable is missing and pushing everything in a dedicated is clearly easier to read and maintain.

I looked at the code and everything looks good from my perspective.
Nothing else to add on my side so feel free to merge when ready.

Many thanks for your help and guidance on this feature.
I learned a lot just from so little code.
